juju/bundle.py

```diff
@@ -156,311 +171,663 @@ def resolve(self, reference):
         """
-        :param charm string:
-            Charm holds the URL of the charm to be added.
+        return "addCharm"
 
-        :param series string:
-            Series holds the series of the charm to be added
-            if the charm default is not sufficient.
+    async def run(self, context):
+        """
+        Run executes a AddCharmChange using the returned parameters from
+        the API server.
+
+        :param context: is used for any methods or properties required to
+            perform a change.
         """
         # We don't add local charms because they've already been added
         # by self._handle_local_charms
-        if charm.startswith('local:'):
-            return charm
+        if self.charm.startswith('local:'):
+            return self.charm
 
-        entity_id = await self.charmstore.entityId(charm)
+        entity_id = await context.charmstore.entityId(self.charm)
         log.debug('Adding %s', entity_id)
-        await self.client_facade.AddCharm(channel=None, url=entity_id, force=False)
+        await context.client_facade.AddCharm(channel=None, url=entity_id, force=False)
         return entity_id
 
-    async def addMachines(self, change_id, params=None):
-        """
-        :param params dict:
-            Dictionary specifying the machine to add. All keys are optional.
-            Keys include:
+    def __str__(self):
+        series = ""
+        channel = ""
+        if self.series != "":
+            series = " for series {}".format(self.series)
+        if self.channel is not None:
+            channel = " from channel {}".format(self.channel)
+        return "upload charm {charm}{series}{channel}".format(charm=self.charm,
+                                                              series=series,
+                                                              channel=channel)
 
-            series: string specifying the machine OS series.
 
-            constraints: string holding machine constraints, if any. We'll
-                parse this into the json friendly dict that the juju api
-                expects.
+class AddMachineChange(ChangeInfo):
+    _toPy = {'series': 'series',
+             'constraints': 'constraints',
+             'container-type': 'container_type',
+             'parent-id': 'parent_id'}
 
-            container_type: string holding the type of the container (for
-                instance ""lxd" or kvm"). It is not specified for top level
-                machines.
+    """
+    AddMachineChange holds a change for adding a machine or container.
 
-            parent_id: string holding a placeholder pointing to another
-                machine change or to a unit change. This value is only
-                specified in the case this machine is a container, in
-                which case also ContainerType is set.
+    :series: holds the optional machine OS series.
+    :constraints: holds the optional machine constraints.
```

Comment:
is this the list of optional params? or just commands in the method not true any more after changes?